clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1075,7 +1084,9 @@ bool PointerVariableConstraint::
         CAtoms::iterator I = vars.begin();
         CAtoms::iterator J = OthCVars.begin();
         while (I != vars.end() && J != OthCVars.end()) {
-          if (CS.getAssignment(*I) != CS.getAssignment(*J)) {
+          ConstAtom *IA = CS.getAssignment(*I);
+          ConstAtom *JA = CS.getAssignment(*J);
+          if (!(IA == CS.getPtr() && JA != CS.getWild()) && IA != JA) {
```

Comment:
Previously, if there were different assignments for I and J then they are considered non-equal (`Ret = false;`). Now non-equal assignments are problematic only when I is not PTR or J is WILD. Why is this?

And, I might rewrite this to be more understandable, if it's correct as is; i.e.,
```
if ((IA != CS.getPtr()) || JA == CS.getWild()) && IA != JA) {
```
I might also reverse the two conjuncts, putting `IA != JA` first, since that's easier to grok WRT the purpose of this function.

Reply:
The reason for this change was that an extra cast was being inserted on calls to `free` because there wasn't an exact match on itypes. The cast insertion pass thinks there needs to be an exact match between the argument and parameter types to use the itype, but this isn't generally the case. I can pass both `_Ptr` and `_Array_ptr` to `free` without inserting a cast.

I've modified how this is handled so it's less likely to affect anything other than the `count(0)` array case.